### DISCUSSION TITLE
Emulating legacy parsers

### DISCUSSION THREAD

**Discussion by dfgordon** [tree-sitter/tree-sitter#1472] [Q&A] [answered]
I am writing an AppleSoft BASIC parser and would like to emulate some "bad behaviors" of the Apple II ROM.  Here is a simple example of 3 lines of code that the legacy parser would parse in exactly the same way (I have verified):

```bas
10 IF X1 THEN 10
20 IFX1THEN10
30 I F X 1 T H E N 1 0
```

In other words it completely ignores spaces.  here is one way I tried to do it:

```js
module.exports = grammar({
	name: 'tinybasic',
	extras: $ => [' '],
	word: $ => $.then_tok,
	rules: {
		source_file: $ => repeat($.line),
		line: $ => seq($.linenum,$.if_tok,$.realvar,$.then_tok,$.linenum,'\n'),
		if_tok: $ => /I *F/,
		then_tok: $ => /T *H *E *N/,
		linenum: $ => /[0-9]([0-9 ]*[0-9])?/,
		realvar: $ => /[A-Z]([A-Z0-9 ]*[A-Z0-9])?/
	}
});
```

Note these regex are allowing runs of spaces in between every character.  What appears to happen is when the engine encounters `X` it eats the rest of the characters on the line as part of the `realvar`.  This is perfectly correct as far as the regex.  I had hoped "context awareness" would allow it to stop eating characters before `THEN`, but no such luck.  Then I thought keyword extraction of `$.then_tok` would get it to subtract out `THEN` before the others, but no luck again.  I have also tried various precedence rules, but nothing could handle each of the three cases thus far.

Is there any way to do this short of external scanners?

**Comment by alemuller** [comment, marked as answer]
I think it is easier to write if you avoid the regex.

```js
const LETTERS = [...'ABCDEFGHIJKLMNOPQRSTUVWXYZ'];
const NUMBERS = [...'0123456789'];

module.exports = grammar({
     name: 'tinybasic',

     extras: $ => [/[ \t]/],

     // FIXME: Tree-sitter bug
     //inline: $ => [ $._eol  ],

     conflicts: $ => [
         [$.variable]
     ],

     rules: {

         program: $ => repeat($._statment),

         _statment: $ => seq(
             $.lineno,
             $._stmt_type,
             $._eol
         ),

         _stmt_type: $ => choice(
             $.if_stmt,
         ),

         // OPTION_1
         // Traditional way to write grammars with tree-sitter.
         //
         // EXCEPTIONALLY for BASIC, you might be better off with the
         // OPTION2 (below, uncomented). 
         //
         //
         // BEGIN CODE OPTION_1
         //if_stmt: $ => seq(
         //    keyword('IF'),
         //    $._expr,
         //    keyword('THEN'),
         //    $._expr
         //),
         // END OPTION_1

         // OPTION_2
         if_stmt: $ => seq(
             alias($._if_tok, 'IF'),
             $._expr,
             alias($._then_tok, 'THEN'),
             $._expr
         ),

         _if_tok: $ => keyword('IF'),

         _then_tok: $ => keyword('THEN'),
         // END OPTION_2

         _expr: $ => choice(
             $._primary,
         ),

         _primary: $ => choice(
             $._number,
             $.variable,
          // $.function
         ),

         _number: $ => choice(
             $.integer
         ),

         variable: $ => seq(
             choice(...LETTERS),
             repeat(choice(
                 ...LETTERS,
                 ...NUMBERS
             )),
         ),

         integer: $ => repeat1(choice(...NUMBERS)),

         lineno: $ => seq(
             choice(...NUMBERS.filter(n => n != '0')),
             repeat(choice(...NUMBERS))
         ),

         _eol: $ => '\n',
     }
})

function keyword(str) {
    return seq(...[...str].map(c=>c))
}
```

`queries/highlights.scm`
```scm
[
 "IF"
 "THEN"
] @keyword

(integer) @number

(variable) @variable

(lineno) @comment
```

To see what nodes they are matching as keyword, use:
```
$ tree-sitter highlight filename.ext
```

I don't know if the codes bellow are legal in BASIC, but tree-sitter can parse them.

```basic
10 IF XTHEN THEN 10
10 IF X THEN THEN 10
```

You can choose which `THEN` will be parsed as variable or keyword using precedence.

**Reply by dfgordon** [reply]
Thanks for the detailed answer!  Both of the suggested options do what I want. So I gather the trick is to use `seq('T','H','E','N')` rather than `/T *H *E *N/`, which I think I can understand.  But why is this done using the interesting notation `seq(...[...str].map(c=>c))`, i.e., why is that different from `seq(...str)`?  I ran this little test program to see if I could notice any difference, but the alert prints out the same thing for both:

```js
str = "THEN";
function test (a,b,c,d) {
    alert(a+b+c+d);
}
test(...[...str].map(c=>c));
test(...str);
```

**Reply by alemuller** [reply]
As you already noticed, javascript can be used to help writing the grammar. But it is not actually included in the grammar.

I don't actually know js, so I can't tell if there is any difference between those fragments. If it compiled and worked, probably not.

The `map` is a "vestigial code" from when I was trying different values of precedence (code below). In other word, I forgot to remove.

Ironically, I thought about either adding "this looks weird" and "map can be used to set precedence" while writing the code.
```js
...[...str].map(c=>prec(number,c))
```

Anyway. Try `prec` with `-1` and `1` with the code bellow and see their difference.

```
10 IF XTHEN THEN 10
10 IF X THEN THEN 10
```

I gave a look on the manual from BASIC. It was written 20 years before LaTeX, so it has that typewriter look (with some ink annotation). It is concise, and enjoyable to read, nothing like moderns specification.

I found it funny how they define the letter "Oh" should be drawn with a slash, but the slashed zero (`0`) became the de facto standard in fixed size fonts.

Just a curiosity. ZSH ignores underscores in options. I wonder if this has something to do with basic.
```zsh
set G_L_O_B
set GL___OB
```

The BASIC expression syntax is also kinda unusual.

**Reply by dfgordon** [reply]
OK that makes sense.  Agreed the old stuff is funny sometimes.  In case anyone is curious I tried
```bas
10 IF X THEN THEN 10
20 IF XTHEN THEN 20
```
on a real Apple II.  This is always parsed the first way (based on `LIST`ing).  However, it does *not* report a syntax error when you `RUN` it (to my surprise).  It simply ignores the statement as far as I can determine.  Trying to set the value of `XTHEN` as in
```bas
10 X THEN = 1
20 XTHEN = 1
```
is again always parsed the first way, but *does* report a syntax error when you `RUN` it.  So unfortunately emulating this parser exactly looks like it may become rather involved.  A compromise might be to just regard all of the above as syntax errors.  I'm not sure exactly how to implement yet, but I think I have enough to figure it out.  

I don't have `tree-sitter highlight` working yet, so will have to try this precedence test later.  Using `tree-sitter parse` did not seem to indicate any difference.

**Comment by dfgordon** [comment]
What I ended up doing is lexing reserved words with regex and identifiers with an external scanner.  Here is why.  The way TS appears to treat regex patterns is, matching starts when context demands it, but once it starts, the matching is context free.  This is perfect for emulating the legacy parser behavior, but it only gets you half-way there.  For the rest, you must *exclude* reserved words from identifiers.  The only way I could see to do that is with the external scanner.  Furthermore, it seems to require the method proposed in #1092.  So this is now all working and I am happy with it.  Interestingly the one case where there is a bit of a discrepancy is the `10 IF X THEN THEN 10` example brought up above.  My TS parser considers that an error, but the Apple II executes that statement silently.

**Reply by alemuller** [reply]
Have you tried using `word`?

See https://tree-sitter.github.io/tree-sitter/creating-parsers#keywords

**Comment by dfgordon** [comment]
This is an interesting point.  Using `word` was my first thought (note code snippet in initial post), but I could not get it to work as desired.  In Applesoft BASIC the reserved words cannot be substrings anywhere in a variable name.  As an example, the parser has to forbid `MYCOS` as a variable name, because `COS` is reserved.  In contrast, what `word` is doing is consuming something like `PRINTA` as an identifier, even though context might tell you this is intended as `seq('PRINT','A')`.

The scanner solution has to verify negative matches against a list of 100 or so reserved words every time you want to lex a variable name, making it perhaps a bit expensive, but so far this has not caused any practical problems.
